[LBSE][GTK][WPE] Composited SVG shapes are blurry when an ancestor is upscaled

obyknovenius opened the pull request:
#### 3d57fb9908dbb3b3cdb6fe1b742c761fc5841df3
<pre>
[LBSE][GTK][WPE] Composited SVG shapes are blurry when an ancestor is upscaled
<a href="https://bugs.webkit.org/show_bug.cgi?id=286468">https://bugs.webkit.org/show_bug.cgi?id=286468</a>

Reviewed by Nikolas Zimmermann.

`GraphicsLayerCoordinated` should calculate and use the root relative scale,
similiarly to what `GraphicsLayerCA` already does.

This patch makes all `svg/compositing` tests pass on GTK and WPE
after slightly updating fuzziness for some of them.

* LayoutTests/platform/glib/TestExpectations:
* LayoutTests/svg/compositing/outermost-svg-directly-composited-group-child.html:
* LayoutTests/svg/compositing/outermost-svg-directly-composited-shape-child.html:
* LayoutTests/svg/compositing/outermost-svg-directly-composited-transformed-group-child.html:
* LayoutTests/svg/compositing/outermost-svg-with-border-overflow-visible.html:
* LayoutTests/svg/compositing/outermost-svg-with-border-padding-margin.html:
* LayoutTests/svg/compositing/outermost-svg-with-border-padding.html:
* LayoutTests/svg/compositing/outermost-svg-with-border.html:
* LayoutTests/svg/compositing/svg-poster-circle.html:
* Source/WebCore/platform/graphics/texmap/coordinated/GraphicsLayerCoordinated.cpp:
(WebCore::GraphicsLayerCoordinated::updateRootRelativeScale):
(WebCore::GraphicsLayerCoordinated::commitLayerChanges):
* Source/WebCore/platform/graphics/texmap/coordinated/GraphicsLayerCoordinated.h:

Canonical link: <a href="https://commits.webkit.org/289615@main">https://commits.webkit.org/289615@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/09f2f645c565235e3212aa1002106248d49cd93b

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/87503 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/7016 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/41878 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/92365 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/38243 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/89554 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/7398 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/15185 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/67583 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/25328 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/90505 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/5640 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/79184 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/47934 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/5426 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/33578 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/37358 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/75838 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/34442 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/94250 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/14668 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/10772 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/76412 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/14922 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/75039 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/75639 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/18607 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/20018 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/18439 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/7611 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/14686 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/19981 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/14429 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/17873 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/16212 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

spenap commented on Source/WebCore/platform/graphics/texmap/coordinated/GraphicsLayerCoordinated.cpp:
```diff
@@ -444,6 +444,39 @@ void GraphicsLayerCoordinated::deviceOrPageScaleFactorChanged()
+    // - only update scale factor when the change is 'significant' (to be defined, (orig - new)/orig > delta?)
+    // - never update the scale factor when it gets smaller (unless we're under memory pressure) (or only periodcally)
+    // - ...
+    // --> For now we disable this logic alltogether, but allow to turn it on selectively (for LBSE)
```
```suggestion
    // - never update the scale factor when it gets smaller (unless we're under memory pressure) (or only periodically)
    // - ...
    // --> For now we disable this logic altogether, but allow to turn it on selectively (for LBSE)
```

obyknovenius replied:
Fixed.

spenap commented on Source/WebCore/platform/graphics/texmap/coordinated/GraphicsLayerCoordinated.h:
```diff
@@ -95,6 +95,10 @@ class GraphicsLayerCoordinated final : public GraphicsLayer {
 
     void deviceOrPageScaleFactorChanged() override;
 
+    float rootRelativeScaleFactor() { return m_rootRelativeScaleFactor; }
```
```suggestion
    float rootRelativeScaleFactor() const { return m_rootRelativeScaleFactor; }
```

nikolaszimmermann replied:
Who needs this accessor, btw?

obyknovenius replied:
A `GraphicsLayerCoordinated` uses its parent's `rootRelativeScaleFactor` to calculate its own:
https://github.com/WebKit/WebKit/pull/39495/files#diff-fd2a9a91ff7f0794e15c1cee3fd7a60b32984fe5a7334bbf02e5efbca2f4d4f0R497

spenap commented on Source/WebCore/platform/graphics/texmap/coordinated/GraphicsLayerCoordinated.cpp:
```diff
@@ -1010,8 +1043,10 @@ void GraphicsLayerCoordinated::commitLayerChanges(CommitState& commitState, floa
     if (m_pendingChanges.contains(Change::ContentsClippingRect))
         m_platformLayer->setContentsClippingRect(m_contentsClippingRect);
 
+    updateRootRelativeScale(); // Needs to happen before Change::ContentsScale.
```
I am not too familiar with this yet, but I understand how we need the "side effect" of the `notLayerPropertyChanged` before we can proceed.

Would it make sense to return `m_rootRelativeScaleFactor` and use the returned value in line 1049, rather than the member variable? I feel that would make the code slightly more robust against refactoring.

nikolaszimmermann replied:
We need to cache the root relative scale factor somewhere, so we can return it if m_shouldUpdateRootRelativeScaleFactor=false, and we want to avoid recomputing it, since it's not cheap.

Or did I misunderstand your proposal?

spenap replied:
Yes, I am OK with the behaviour inside the function (`updateRootRelativeScale()`) itself, so I agree with caching the factor in a member variable.

My point is that rather than _using_ the member variable in this function, which means we 1) need to call the function that updates it 2) we need to use the member, and we have an _implicit_ linear dependency between the two (e.g. "before you can use this property, update it"), I think it's better if we instead do `auto scale = updateRootRelativeScale()`, then use `scale`. 

As of today, there would be no functional difference. But in an hypothetical refactoring down the line, having scale into a member variable initialised by the method is clearer, as they two need to remain tied together.

I hope it makes sense :-)

nikolaszimmermann commented on LayoutTests/platform/glib/TestExpectations:
```diff
@@ -1930,10 +1930,18 @@ imported/w3c/web-platform-tests/svg/animations/reinserting-svg-into-document.htm
+svg/compositing/outermost-svg-with-border-padding-margin.html  [ ImageOnlyFailure ]
+svg/compositing/outermost-svg-with-border-padding.html  [ ImageOnlyFailure ]
+svg/compositing/outermost-svg-with-border.html  [ ImageOnlyFailure ]
+svg/compositing/svg-poster-circle.html  [ ImageOnlyFailure ]
```
Would be great if you could comment the test results, at least in the PR. The commit message also doesn't mention yet that you unskipped tests -- would be natural to explain the changes there.

obyknovenius replied:
I added a comment regarding these tests.

obyknovenius committed:
[LBSE][GTK][WPE] Composited SVG shapes are blurry when an ancestor is upscaled
https://bugs.webkit.org/show_bug.cgi?id=286468

Reviewed by Nikolas Zimmermann.

`GraphicsLayerCoordinated` should calculate and use the root relative scale,
similiarly to what `GraphicsLayerCA` already does.

This patch makes all `svg/compositing` tests pass on GTK and WPE
after slightly updating fuzziness for some of them.

* LayoutTests/platform/glib/TestExpectations:
* LayoutTests/svg/compositing/outermost-svg-directly-composited-group-child.html:
* LayoutTests/svg/compositing/outermost-svg-directly-composited-shape-child.html:
* LayoutTests/svg/compositing/outermost-svg-directly-composited-transformed-group-child.html:
* LayoutTests/svg/compositing/outermost-svg-with-border-overflow-visible.html:
* LayoutTests/svg/compositing/outermost-svg-with-border-padding-margin.html:
* LayoutTests/svg/compositing/outermost-svg-with-border-padding.html:
* LayoutTests/svg/compositing/outermost-svg-with-border.html:
* LayoutTests/svg/compositing/svg-poster-circle.html:
* Source/WebCore/platform/graphics/texmap/coordinated/GraphicsLayerCoordinated.cpp:
(WebCore::GraphicsLayerCoordinated::updateRootRelativeScale):
(WebCore::GraphicsLayerCoordinated::commitLayerChanges):
* Source/WebCore/platform/graphics/texmap/coordinated/GraphicsLayerCoordinated.h:

Canonical link: https://commits.webkit.org/289615@main
